File: week1/1-3.cpp
/* [21년 재직자 대회 예선] 좌석 관리 */

#include <iostream>
#include <vector>
#include <map>
#include <limits>
#include <cmath>

using namespace std;

vector<vector<bool>> isSeatOccupied; // false : not in use, true : in use
map<int, int> idStatus; // 0 : not eat, 1 : eating, 2 : finished eating
map<int, vector<int>> idLocation;

float getNearestDistance(int x, int y){
    float minDistance = numeric_limits<float>::max();

    int iterLen = idLocation.size();
    for(const auto& pair : idLocation){
        const std::vector<int>& location = pair.second;

        float distance = sqrt(pow(x-location[0], 2) + pow(y-location[1], 2));
        if(distance < minDistance){
            minDistance = distance;
        }
    }

    return minDistance;
}

bool assignSeat(int id, int N, int M){
    float targetDistance = 0.0;
    int target_x = 0;
    int target_y = 0;

    for(int i = 1; i <= N; i++){
        for(int j = 1; j <= M; j++){
            // continue if the seat or adjacent seats are occupied
            if(isSeatOccupied[i][j] || isSeatOccupied[i][j-1] || isSeatOccupied[i][j+1] || isSeatOccupied[i-1][j] || isSeatOccupied[i+1][j]){
                continue;
            }

            float currentDistance = getNearestDistance(i, j);
            if(targetDistance < currentDistance){
                targetDistance = currentDistance;
                target_x = i;
                target_y = j;
            }
        }
    }

    // if there are no available seats
    if(!idLocation.empty() && targetDistance == 0.0){
        return false;
    }

    idStatus[id] = 1;
    idLocation[id] = {target_x, target_y};
    isSeatOccupied[target_x][target_y] = true;

    return true;
}

void leaveSeat(int id){
    int x = idLocation[id][0];
    int y = idLocation[id][1];

    idStatus[id] = 2;
    isSeatOccupied[x][y] = false;
    idLocation.erase(id);
    cout << id << " leaves from the seat (" << x << ", " << y << ")." << endl;

    return;
}

int main(){
    int N, M, Q;
    cin >> N >> M >> Q;

    isSeatOccupied.resize(N+2, vector<bool>(M+2, false)); // with margins

    for(int i = 0; i < Q; i++){
        string operation;
        int id;
        cin >> operation >> id;

        if(operation == "In"){
            if(idStatus[id] == 0){ // assign a seat
                if(assignSeat(id, N, M)){
                    cout << id << " gets the seat (" << idLocation[id][0] << ", " << idLocation[id][1] << ")." << endl;
                }
                else{
                    cout << "There are no more seats." << endl;
                }
            }
            else if(idStatus[id] == 1){ // already seated
                cout << id << " already seated." << endl;
            }
            else{ // already ate lunch
                cout << id << " already ate lunch." << endl;
            }
        }
        else{
            if((idStatus.find(id) == idStatus.end()) || (idStatus[id] == 0)){ // didn't eat lunch
                cout << id << " didn't eat lunch." << endl;
            }
            else if(idStatus[id] == 1){ // leaves from the seat
                leaveSeat(id);
            }
            else{ // already left seat
                cout << id << " already left seat." << endl;
            }
        }
    }

    return 0;
}

File: final/1-3.py
# [21년 재직자 대회 예선] 좌석 관리
import sys
input = sys.stdin.readline

def get_nearest_distance(x,y):
    min_distance = sys.maxsize
    
    for location in id_location:
        distance = ((x-location[0])**2 + (y-location[1]**2))**(1/2)
        if(distance < min_distance):
            min_distance = distance
            
    return min_distance

def assign_seat(id):
    distance = 0
    x,y = 0,0
    
    for i in range(1,N+1):
        for j in range(1,M+1):
            if(cafeteria[i][j] or cafeteria[i-1][j] or cafeteria[i+1][j] or cafeteria[i][j-1] or cafeteria[i][j+1]): # 해당 좌석 혹은 상하좌우의 좌석이 이미 선점되었다면 넘어감
                continue
            
            current_distance = get_nearest_distance(i,j)
            if(distance < current_distance):
                distance = current_distance
                x,y = i,j
    
    if(distance == 0 and id_location):
        return False
    
    id_status[id] = 1 # 식사 중으로 표시
    cafeteria[x][y] = True # 자리 할당
    id_location[id] = [x,y] # 자리 할당
    return True

def leave_seat(id,x,y):
    id_status[id] = 2 # 식사 완료로 표시
    cafeteria[x][y] = False # 자리 비우기
    del(id_location[id]) # 자리 비우기

# main
N,M,Q = map(int, input().split())
cafeteria = [[False for _ in range(M+2)] for _ in range(N+2)] # NxM + 상하좌우 padding

id_status = dict() # key: id, value: status (0: 식사 X, 1: 식사 중, 2: 식사 완료)
id_location = dict() # key: id, value: [x,y] (현재 앉은 위치)

for _ in range(Q):
    task = input()
    
    if(task[:2] == "In"): # In
        id = task[3]
        
        if(id_status[id] == 1): # {id} already seated.
            print(f"{id} already seated.")
        elif(id_status[id] == 2): # {id} already ate lunch.
            print(f"{id} already ate lunch.")
        else:
            if(assign_seat() == True): # 배정 성공
                x,y = id_location[id]
                print(f"{id} gets the seat ({x}, {y}).")
            else: # 배정 실패
                print("There are no more seats.")
        
    else: # Out
        id = task[4]
        
        if(id_status[id] == 0): # {id} didn't eat lunch. -> not in 등으로 고치기
            print(f"{id} didn't eat lunch.")
        elif(id_status[id] == 2): # {id} already left seat.
            print(f"{id} already left seat.")
        else: # {id} leaves from the seat ({x},{y}).
            x,y = id_location[id]
            print(f"{id} leaves from the seat ({x}, {y}).")
            leave_seat(id,x,y)